File: 55.jump-game/greedy.rb
#
# @lc app=leetcode id=55 lang=ruby
#
# [55] Jump Game
#

# @lc code=start
# @param {Integer[]} nums
# @return {Boolean}
def can_jump(nums)
  last_possible_position = nums.size - 1

  (nums.size-1).downto(0) do |i|
    last_possible_position = i if i + nums[i] >= last_possible_position
  end

  last_possible_position == 0
end
# @lc code=end

File: 55.jump-game/greedy.js
/*
 * @lc app=leetcode.cn id=55 lang=javascript
 * @lcpr version=30204
 *
 * [55] 跳跃游戏
 */


// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
  if (nums.length === 1) return true;

  let maxReach = 0, i = 0;
  while (i <= maxReach) {
    reach = i + nums[i];
    if (reach >= nums.length - 1) return true;

    maxReach = Math.max(maxReach, reach);
    i++;
  }

  return false;
};
// @lc code=end



// @lcpr case=start
// [2,3,1,1,4]\n
// @lcpr case=end

// @lcpr case=start
// [3,2,1,0,4]\n
// @lcpr case=end

// @lcpr case=start
// [3]\n
// @lcpr case=end

// @lcpr case=start
// [0]\n
// @lcpr case=end

// @lcpr case=start
// [3,2,0,1,2,0,4,1,2,2,1,3,0,2,0]\n
// @lcpr case=end

// @lcpr case=start
// [0,2,3]\n
// @lcpr case=end
